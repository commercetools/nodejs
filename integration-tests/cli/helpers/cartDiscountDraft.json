{
  "name": {
    "en": "integration-test-cart-discount"
  },
  "value": {
    "type": "relative",
    "permyriad": 1000
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "target": {
    "type": "lineItems",
    "predicate": "1 = 1"
  },
  "sortOrder": "0.1",
  "requiresDiscountCode": true
}
