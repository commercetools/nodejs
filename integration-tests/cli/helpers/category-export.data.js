export default [
  {
    key: 'fooKey',
    name: {
      en: 'categoryA',
    },
    slug: {
      en: 'category-a',
    },
    description: {
      en: 'descA',
    },
    externalId: 'a',
    assets: [],
  },
  {
    key: 'barKey',
    name: {
      en: 'categoryB',
    },
    slug: {
      en: 'category-b',
    },
    description: {
      en: 'descB',
    },
    externalId: 'b',
    assets: [],
  },
  {
    key: 'fooBarKey',
    name: {
      en: 'categoryC',
    },
    slug: {
      en: 'category-c',
    },
    description: {
      en: 'descC',
    },
    externalId: 'c',
    assets: [],
  },
]
