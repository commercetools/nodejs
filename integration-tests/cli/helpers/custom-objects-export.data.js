export default [
  {
    container: 'Ludus',
    key: 'copperKey',
    value: {
      paymentMethod: 'Cash',
      paymentID: '1',
    },
  },
  {
    container: 'Frobozz',
    key: 'jadeKey',
    value: {
      paymentMethod: 'cc',
      paymentID: '2',
    },
  },
  {
    container: 'Syrinx',
    key: 'crystalKey',
    value: {
      paymentMethod: 'new',
      paymentID: '3',
    },
  },
]
