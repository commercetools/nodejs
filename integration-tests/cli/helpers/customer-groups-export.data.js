export default [
  {
    groupName: 'Ludus',
    key: 'copperKey',
  },
  {
    groupName: 'Frobozz',
  },
  {
    groupName: 'Syrinx',
    key: 'crystalKey',
  },
]
