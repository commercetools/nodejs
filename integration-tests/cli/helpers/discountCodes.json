[{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "IT8MKKZX"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlN85n0"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITDHzsVX"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlb6Z1u"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITfWgUVd"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITpQgE1P"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITL2pvbZ"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlrnkj5"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITDJO3LT"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITuIcDI2"
}]
