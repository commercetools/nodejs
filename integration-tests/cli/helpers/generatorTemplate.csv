name.en,name.de,description.en,description.de,cartPredicate,isActive,maxApplications,maxApplicationsPerCustomer
Sammy,Valerian,greatest promo,super angebot,some cart predicate,true,9,4
