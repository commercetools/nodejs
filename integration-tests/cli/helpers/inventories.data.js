export const inventories = [{
  sku: '12345',
  quantityOnStock: 20,
  custom: {
    type: {
      key: 'custom-type',
    },
    fields: {
      description: 'integration tests!! arrgggh',
    },
  },
}]

export const customFields = [{
  key: 'custom-type',
  name: {
    en: 'customized fields',
  },
  description: {
    en: 'custom description',
  },
  resourceTypeIds: ['inventory-entry'],
  fieldDefinitions: [{
    name: 'description',
    type: {
      name: 'String',
    },
    required: false,
    label: {
      en: 'owner',
    },
    inputHint: 'SingleLine',
  }],
}]
