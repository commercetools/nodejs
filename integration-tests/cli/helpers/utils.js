import { createAuthMiddlewareForClientCredentialsFlow } from '@commercetools/sdk-middleware-auth'
import { createClient } from '@commercetools/sdk-client'
import { createRequestBuilder } from '@commercetools/api-request-builder'
import { createHttpMiddleware } from '@commercetools/sdk-middleware-http'
import fetch from 'node-fetch'

export function clearData(apiConfig, entityName) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })

  const service = createRequestBuilder({
    projectKey: apiConfig.projectKey,
  })[entityName]

  const request = {
    uri: service.build(),
    method: 'GET',
  }
  return client.process(request, payload => {
    // Built-in states cannot be deleted
    const results =
      entityName === 'states'
        ? payload.body.results.filter(state => state.builtIn === false)
        : payload.body.results
    return Promise.all(
      results.map(result =>
        client.execute({
          uri: service
            .byId(result.id)
            .withVersion(result.version)
            .build(),
          method: 'DELETE',
        })
      )
    )
  })
}

export function createData(apiConfig, entityName, data, id) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })
  const requestOption = { projectKey: apiConfig.projectKey }
  const service = createRequestBuilder(requestOption)[entityName]
  return Promise.all(
    data.map(_data => {
      if (id) service.byId(id)
      const request = {
        uri: service.build(),
        method: 'POST',
        body: _data,
      }
      return client.execute(request)
    })
  )
}

export function getId(apiConfig, entityName) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })

  const service = createRequestBuilder({
    projectKey: apiConfig.projectKey,
  })[entityName]

  const request = {
    uri: service.build(),
    method: 'GET',
  }

  return client
    .execute(request)
    .then(result => Promise.resolve(result.body.results[0].id))
}
