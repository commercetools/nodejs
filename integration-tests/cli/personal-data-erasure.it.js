import PersonalDataErasure from '@commercetools/personal-data-erasure'
import tmp from 'tmp'
import { getCredentials } from '@commercetools/get-credentials'
import { exec } from 'mz/child_process'
import { version } from '@commercetools/personal-data-erasure/package.json'
import { createData, clearData, getId } from './helpers/utils'
import {
  customer,
  order,
  cart,
  payment,
  shoppingList,
  review,
  customLineItem,
} from './helpers/personal-data-erasure.data'

let projectKey
if (process.env.CI === 'true') projectKey = 'custom-objects-import-int-tests'
else projectKey = process.env.npm_config_projectkey

describe('personal data erasure', () => {
  jest.setTimeout(15000) // 15 second timeout

  const bin = './integration-tests/node_modules/.bin/personal-data-erasure'

  describe('CLI basic functionality', () => {
    it('should print usage information given the help flag', async () => {
      const [stdout, stderr] = await exec(`${bin} --help`)
      expect(stderr).toBeFalsy()
      expect(stdout).toMatchSnapshot()
    })

    it('should print the module version given the version flag', async () => {
      const [stdout, stderr] = await exec(`${bin} --version`)
      expect(stderr).toBeFalsy()
      expect(stdout).toBe(`${version}\n`)
    })
  })

  // FIXME: currently it fails with following error.
  // CustomLineItem error: The referenced object of type 'tax-category' with identifier 'd205cc42-e399-424a-b6fc-0ef44772d6bc' was not found. It either doesn't exist, or it can't be accessed from this endpoint (e.g., if the endpoint filters by store or customer account).
  describe.skip('normal usage', () => {
    let apiConfig
    let personalDataErasure
    let customerId
    let cartId

    const logger = {
      error: () => {},
      warn: () => {},
      info: () => {},
      debug: () => {},
    }

    const setupProject = async () => {
      const credentials = await getCredentials(projectKey)
      apiConfig = {
        host: 'https://auth.europe-west1.gcp.commercetools.com',
        apiUrl: 'https://api.europe-west1.gcp.commercetools.com',
        projectKey,
        credentials,
      }
      // Clear all existing data
      await Promise.all([
        clearData(apiConfig, 'customers'),
        clearData(apiConfig, 'orders'),
        clearData(apiConfig, 'carts'),
        clearData(apiConfig, 'payments'),
        clearData(apiConfig, 'shoppingLists'),
        clearData(apiConfig, 'reviews'),
      ])

      await createData(apiConfig, 'customers', customer)
      customerId = await getId(apiConfig, 'customers')
      await createData(apiConfig, 'carts', [{ ...cart[0], customerId }])
      cartId = await getId(apiConfig, 'carts')

      await createData(apiConfig, 'carts', customLineItem, cartId)

      await Promise.all([
        createData(apiConfig, 'orders', [{ ...order[0], id: cartId }]),
        createData(apiConfig, 'payments', [
          { ...payment[0], customer: { id: customerId, typeId: 'customer' } },
        ]),
        createData(apiConfig, 'shoppingLists', [
          {
            ...shoppingList[0],
            customer: { id: customerId, typeId: 'customer' },
          },
        ]),
        createData(apiConfig, 'reviews', [
          { ...review[0], customer: { id: customerId, typeId: 'customer' } },
        ]),
      ])
    }

    beforeEach(() => {
      personalDataErasure = new PersonalDataErasure({ apiConfig }, logger)
    })

    describe('functional usage', () => {
      beforeAll(async () => {
        await setupProject()
      }, 30000) // 30 second timeout as this might take a while

      describe('::getCustomerData', () => {
        it('should get data on the CTP', async () => {
          const data = await personalDataErasure.getCustomerData(customerId)

          expect(data).toHaveLength(4) // new data length is 4
          expect(data).toContainEqual(
            expect.objectContaining({ type: 'CartCreated' })
          )
          expect(data).toContainEqual(
            expect.objectContaining({ type: 'PaymentCreated' })
          )
          expect(data).toContainEqual(
            expect.objectContaining({ type: 'CustomerCreated' })
          )
          expect(data).toContainEqual(
            expect.objectContaining({ type: 'ReviewCreated' })
          )
          expect(data).toContainEqual(
            expect.objectContaining({ type: 'OrderCreated' })
          )
          expect(data).toContainEqual(
            expect.objectContaining({ type: 'CartCreated' })
          )
          expect(data).toContainEqual(
            expect.objectContaining({ type: 'Order' })
          )
          expect(data).toContainEqual(expect.objectContaining({ type: 'Cart' }))
          expect(data).toContainEqual(
            expect.objectContaining({ email: 'foo@bar.de' })
          )
          expect(data).toContainEqual(
            expect.objectContaining({
              amountPlanned: {
                type: 'centPrecision',
                currencyCode: 'EUR',
                centAmount: 100,
                fractionDigits: 2,
              },
            })
          )
          expect(data).toContainEqual(
            expect.objectContaining({
              name: { de: 'deutscherListenName', en: 'englishListName' },
            })
          )
          expect(data).toContainEqual(
            expect.objectContaining({ text: 'Review text' })
          )
        })
      })

      describe('::deleteAll', () => {
        it('should delete data on the CTP', async () => {
          let data
          const fetchDataAfterTimeout = (ms) =>
            new Promise((resolve) =>
              setTimeout(async () => {
                data = await personalDataErasure.getCustomerData(customerId)
                resolve()
              }, ms)
            )

          await personalDataErasure.deleteAll(customerId)

          // wait 3s for DB to finish deletion
          await fetchDataAfterTimeout(3000)
          expect(data).toHaveLength(0)
        })
      })
    })

    describe('child process usage', () => {
      beforeAll(async () => {
        await setupProject()
      }, 30000) // 30 second timeout as this might take a while

      describe('get data', () => {
        it('should log success messages', async () => {
          const filePath = tmp.fileSync().name
          const [stdout, stderr] = await exec(
            `${bin} -o ${filePath} -p ${projectKey} -c ${customerId}`
          )
          expect(stderr).toBeFalsy()
          expect(stdout).toMatch(/Starting to fetch data/)
          expect(stdout).toMatch(/Export operation completed successfully/)
          expect(stdout).toMatch(
            /entities has been successfully written to file/
          )
        })
      })

      describe('delete data', () => {
        it('should log success messages', async () => {
          const filePath = tmp.fileSync().name
          const [stdout, stderr] = await exec(
            `${bin} -o ${filePath} -p ${projectKey} -c ${customerId} -D  --force`
          )
          expect(stderr).toBeFalsy()
          expect(stdout).toMatch(/Starting deletion/)
          expect(stdout).toMatch(
            `All data related to customer with id '${customerId}' has successfully been deleted.`
          )
        })
      })
    })
  })
})
