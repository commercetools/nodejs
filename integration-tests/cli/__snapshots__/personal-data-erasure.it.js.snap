// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`personal data erasure CLI basic functionality should print usage information given the help flag 1`] = `
"
Usage: personal-data-erasure [options]
Export and delete all data related to a single customer

Options:
  --help            Show help text.                                    [boolean]
  --version         Show version number                                [boolean]
  --output, -o      Path to output file.                     [default: \\"stdout\\"]
  --customerId, -c  Customer to fetch or delete.                      [required]
  --apiUrl          The host URL of the HTTP API service.
                                              [default: \\"https://api.sphere.io\\"]
  --authUrl         The host URL of the OAuth API service.
                                             [default: \\"https://auth.sphere.io\\"]
  --accessToken     CTP client access token.
  --projectKey, -p  API project key.                                  [required]
  --deleteAll, -D   Delete all data related to customer.               [boolean]
  --force           Continue without confirmation when combined with
                    --deleteAll.                                       [boolean]
  --logLevel        Logging level: error, warn, info or debug. [default: \\"info\\"]
  --prettyLogs      Pretty print logs to the terminal                  [boolean]
  --logFile         Path to where to save logs file.
                                 [string] [default: \\"personal-data-erasure.log\\"]

"
`;
